// @blockstack/ui/dist/codeblock/index.esm.js
import { objectWithoutPropertiesLoose as _objectWithoutPropertiesLoose, objectSpread2 as _objectSpread2 } from '../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import { Box } from '../box/index.esm.js';
import { Highlighter } from '../highlighter/index.esm.js';

var CodeBlock = /*#__PURE__*/React__default.forwardRef(function (_ref, ref) {
  var code = _ref.code,
      showLineNumbers = _ref.showLineNumbers,
      hideLineHover = _ref.hideLineHover,
      _ref$style = _ref.style,
      style = _ref$style === void 0 ? {} : _ref$style,
      language = _ref.language,
      Prism = _ref.Prism,
      rest = _objectWithoutPropertiesLoose(_ref, ["code", "showLineNumbers", "hideLineHover", "style", "language", "Prism"]);

  return React__default.createElement(Box, Object.assign({
    overflowX: "auto",
    bg: "ink",
    borderRadius: [0, 0, '12px'],
    py: "base",
    width: "100%",
    ref: ref,
    style: _objectSpread2({}, style, {
      whiteSpace: 'pre',
      fontFamily: 'Fira Code, Consolata, monospace',
      fontSize: '14px'
    })
  }, rest), React__default.createElement(Highlighter, {
    language: language,
    code: code.toString().trim(),
    showLineNumbers: showLineNumbers,
    hideLineHover: hideLineHover,
    Prism: Prism
  }));
});

export { CodeBlock };
