// @stacks/network/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StacksRegtest = exports.StacksMocknet = exports.StacksTestnet = exports.StacksMainnet = exports.HIRO_MOCKNET_DEFAULT = exports.HIRO_TESTNET_DEFAULT = exports.HIRO_REGTEST_DEFAULT = exports.HIRO_MAINNET_DEFAULT = void 0;
const common_1 = require("@stacks/common");
exports.HIRO_MAINNET_DEFAULT = 'https://stacks-node-api.mainnet.stacks.co';
exports.HIRO_REGTEST_DEFAULT = 'https://stacks-node-api.regtest.stacks.co';
exports.HIRO_TESTNET_DEFAULT = 'https://stacks-node-api.testnet.stacks.co';
exports.HIRO_MOCKNET_DEFAULT = 'http://localhost:3999';
class StacksMainnet {
    constructor(networkUrl = { url: exports.HIRO_MAINNET_DEFAULT }) {
        this.version = common_1.TransactionVersion.Mainnet;
        this.chainId = common_1.ChainID.Mainnet;
        this.bnsLookupUrl = 'https://stacks-node-api.mainnet.stacks.co';
        this.broadcastEndpoint = '/v2/transactions';
        this.transferFeeEstimateEndpoint = '/v2/fees/transfer';
        this.accountEndpoint = '/v2/accounts';
        this.contractAbiEndpoint = '/v2/contracts/interface';
        this.readOnlyFunctionCallEndpoint = '/v2/contracts/call-read';
        this.isMainnet = () => this.version === common_1.TransactionVersion.Mainnet;
        this.getBroadcastApiUrl = () => `${this.coreApiUrl}${this.broadcastEndpoint}`;
        this.getTransferFeeEstimateApiUrl = () => `${this.coreApiUrl}${this.transferFeeEstimateEndpoint}`;
        this.getAccountApiUrl = (address) => `${this.coreApiUrl}${this.accountEndpoint}/${address}?proof=0`;
        this.getAbiApiUrl = (address, contract) => `${this.coreApiUrl}${this.contractAbiEndpoint}/${address}/${contract}`;
        this.getReadOnlyFunctionCallApiUrl = (contractAddress, contractName, functionName) => `${this.coreApiUrl}${this.readOnlyFunctionCallEndpoint}/${contractAddress}/${contractName}/${encodeURIComponent(functionName)}`;
        this.getInfoUrl = () => `${this.coreApiUrl}/v2/info`;
        this.getBlockTimeInfoUrl = () => `${this.coreApiUrl}/extended/v1/info/network_block_times`;
        this.getPoxInfoUrl = () => `${this.coreApiUrl}/v2/pox`;
        this.getRewardsUrl = (address, options) => {
            let url = `${this.coreApiUrl}/extended/v1/burnchain/rewards/${address}`;
            if (options) {
                url = `${url}?limit=${options.limit}&offset=${options.offset}`;
            }
            return url;
        };
        this.getRewardsTotalUrl = (address) => `${this.coreApiUrl}/extended/v1/burnchain/rewards/${address}/total`;
        this.getRewardHoldersUrl = (address, options) => {
            let url = `${this.coreApiUrl}/extended/v1/burnchain/reward_slot_holders/${address}`;
            if (options) {
                url = `${url}?limit=${options.limit}&offset=${options.offset}`;
            }
            return url;
        };
        this.getStackerInfoUrl = (contractAddress, contractName) => `${this.coreApiUrl}${this.readOnlyFunctionCallEndpoint}
    ${contractAddress}/${contractName}/get-stacker-info`;
        this._coreApiUrl = networkUrl.url;
    }
    get coreApiUrl() {
        return this._coreApiUrl;
    }
    set coreApiUrl(_url) {
        throw new Error('Cannot modify property `coreApiUrl` after object initialization');
    }
    getNameInfo(fullyQualifiedName) {
        const nameLookupURL = `${this.bnsLookupUrl}/v1/names/${fullyQualifiedName}`;
        return common_1.fetchPrivate(nameLookupURL)
            .then(resp => {
            if (resp.status === 404) {
                throw new Error('Name not found');
            }
            else if (resp.status !== 200) {
                throw new Error(`Bad response status: ${resp.status}`);
            }
            else {
                return resp.json();
            }
        })
            .then(nameInfo => {
            if (nameInfo.address) {
                return Object.assign({}, nameInfo, { address: nameInfo.address });
            }
            else {
                return nameInfo;
            }
        });
    }
}
exports.StacksMainnet = StacksMainnet;
class StacksTestnet extends StacksMainnet {
    constructor(networkUrl = { url: exports.HIRO_TESTNET_DEFAULT }) {
        super(networkUrl);
        this.version = common_1.TransactionVersion.Testnet;
        this.chainId = common_1.ChainID.Testnet;
    }
}
exports.StacksTestnet = StacksTestnet;
class StacksMocknet extends StacksMainnet {
    constructor(networkUrl = { url: exports.HIRO_MOCKNET_DEFAULT }) {
        super(networkUrl);
        this.version = common_1.TransactionVersion.Testnet;
        this.chainId = common_1.ChainID.Testnet;
    }
}
exports.StacksMocknet = StacksMocknet;
class StacksRegtest extends StacksMainnet {
    constructor(networkUrl = { url: exports.HIRO_REGTEST_DEFAULT }) {
        super(networkUrl);
        this.version = common_1.TransactionVersion.Testnet;
        this.chainId = common_1.ChainID.Testnet;
    }
}
exports.StacksRegtest = StacksRegtest;
