// @blockstack/ui/dist/toast/toaster.esm.js
import { objectWithoutPropertiesLoose as _objectWithoutPropertiesLoose } from '../_virtual/_rollupPluginBabelHelpers.js';
import React__default, { useCallback } from 'react';
import { Box } from '../box/index.esm.js';
import { Flex } from '../flex/index.esm.js';
import { Toast } from './toast.esm.js';

var Toaster = function Toaster(_ref) {
  var toasts = _ref.toasts,
      removeToast = _ref.removeToast,
      rest = _objectWithoutPropertiesLoose(_ref, ["toasts", "removeToast"]);

  var onClear = useCallback(function (id) {
    removeToast(id);
  }, [removeToast]);
  return React__default.createElement(Flex, Object.assign({
    align: "center",
    justify: "flex-end",
    flexDirection: "column",
    position: "fixed",
    width: "100%",
    height: "100vh",
    bottom: 0,
    zIndex: 9999999,
    style: {
      pointerEvents: 'none'
    }
  }, rest), toasts.map(function (_ref2) {
    var id = _ref2.id,
        rest = _objectWithoutPropertiesLoose(_ref2, ["id"]);

    return React__default.createElement(Box, {
      key: id
    }, React__default.createElement(Toast, Object.assign({
      id: id,
      onClear: onClear
    }, rest)));
  }));
};

export { Toaster };
